src/test/scala/com/mozilla/telemetry/sinks/HttpSinkMetricsTest.scala

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package com.mozilla.telemetry.sinks
+
+import java.io.File
+
+import org.apache.spark.SparkEnv
+import org.apache.spark.metrics.source.custom.{AccumulatorMetricsSource, AccumulatorMetricsSourceTestBase}
+import org.apache.spark.sql.SparkSession
+import org.scalatest.{BeforeAndAfterEach, FlatSpec, Matchers, PrivateMethodTester}
+
+import scala.io.Source
+
+class HttpSinkMetricsTest extends FlatSpec
+  with AccumulatorMetricsSourceTestBase with HttpSinkTestBase
+  with BeforeAndAfterEach
+  with Matchers with PrivateMethodTester {
+
+  override protected val MetricsSourceName = "HttpSink"
+  override protected val MetricName = "success"
+  val NumAccumulators = 5
+
+  var metricsSource: AccumulatorMetricsSource = _
+  var spark: SparkSession = _
+
+  override protected def beforeEach(): Unit = {
+    metricsSource = new AccumulatorMetricsSource("HttpSink")
+
+    spark = SparkSession.builder()
+      .appName(appName)
+      .master("local[2]")
+      .config("spark.metrics.conf.*.sink.csv.class", "org.apache.spark.metrics.sink.CsvSink")
+      .config("spark.metrics.conf.*.sink.csv.period", "1")
+      .config("spark.metrics.conf.*.sink.csv.directory", MetricsOutputDir)
+      .config("spark.metrics.namespace", "${spark.app.name}")
+      .getOrCreate()
+
+    val httpSinkConfig = HttpSink.Config(
+      maxAttempts = maxAttempts,
+      defaultDelayMillis = delay,
+      connectionTimeoutMillis = timeout)
+      .withMetrics(metricsSource)(spark.sparkContext)
+
+    httpSink = AmplitudeHttpSink(
+      apiKey,
+      s"http://$Host:$Port$Path",
+      httpSinkConfig)
+
+    super.beforeEach()
+  }
+
+  override protected def afterEach(): Unit = {
+    try super.afterEach()
+    finally spark.stop()
+  }
+
+  "HttpSink.Metrics" should "output metrics" in {
```

Comment:
I'm wondering if this test doesn't go too far with mixing testing of metrics source (which is already covered elsewhere - there's some copy-pasta from base accumulator metric test) and http sink behavior.

The new part it's testing (that hasn't been tested already) is accumulator usage in HttpSink (whether they're incremented correctly). Maybe cleaner way for testing this would be to only check accumulator values in the end?

Reply:
Good point. I agree checking just accumulator values should be sufficient here.